test/cpp/tensorexpr/test_simplify.cpp

```diff
@@ -2609,6 +2734,164 @@ void testSimplifyDivisionScalarFactorization() {
   }
 }
 
+void testSimplifyDivisionSymbolicFactorization() {
+  KernelScope kernel_scope;
+
+  {
+    // Same numerator and denomenator still cancel.
+    // x / x => 1.
+    VarHandle x("x", kInt);
+    ExprHandle body = (x * 8) / (x * 4);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    IS_IMM_WITH_VAL(Int, simplified.node(), 2);
+  }
+
+  {
+    // Simple factorization of numerator and denominator with different scalars.
+    // 8x / 4x => 2.
```

Comment:
The same concern as @ZolotukhinM pointed out above applies here and in several cases below as well, right? What if `x` can be `0`? Is it okay to simplify such terms in that case?

Reply:
I agree for the `Mod` case since it's reasonable to have zero on the LHS of a Mod (not the RHS). For `Div` I think it's fine to assume the user is not intentionally dividing by zero, or at least if they are they don't care about the result.

Reply:
> For Div I think it's fine to assume the user is not intentionally dividing by zero, or at least if they are they don't care about the result.

I'm not entirely sure about this. Imagine I have `a[i,j] / a[i,j]`: if there are 0 elements in `a[i,j]` then it'll be either an exception or NaN. If we optimize it to 1, we change the model's behavior.

Reply:
Yeah, it'll be NaN:
```
>>> a = torch.tensor([0, 1, 2], dtype=torch.int32)
>>> b = torch.div(a, a)
>>> b
tensor([nan, 1., 1.])
```

Reply:
I'm trying to think of an actual situation that's important to preserve. Feels a bit like cutting off our nose to spite our face if we prevent an optimization because it might prevent a crash that should occur. Depends if you regard the crash as an important signal I guess.

This would be fixable with a separate zero-check I guess, but that defeats the point of simpliying.

Reply:
I don't think we have a choice here. Since eager mode is our reference, we need to follow what happens there and thus such simplification is not allowed. Note that this example could be rewritten in a probably more convincing way:
```
import torch

def test(x, y):
    return x*y/y

scripted = torch.jit.script(test)
eager_result = test(torch.zeros(1, device='cuda'), torch.zeros(1, device='cuda'))
# Warmup
for _ in range(10):
    scripted(torch.zeros(1, device='cuda'), torch.zeros(1, device='cuda'))

# Run
script_result = scripted(torch.zeros(1, device='cuda'), torch.zeros(1, device='cuda'))

assert_all_close(eager_result, script_result)
```

Reply:
You guys are right, i'll remove the symbolic Div simplification.

Reply:
Why can't check whether the divisor is zero in the lowering, before attempting the division? This would still allow the simplification in other cases.

Reply:
How would a separate zero-check defeat the point of simplifying? We would still simplify what ends up being generated on the non-zero divisor branch, which is still something.

Reply:
But that would require adding a conditional. Is it worth simplifying at all in that case?

Reply:
I don't see why it wouldn't be worth it. Such expressions could be generated from contexts which aren't an `aten::div` and in which the divisor is known not to be zero. Also, a conditional isn't that expensive compared to what's being simplified here.

In general, I don't think we should tie the semantics and the opportunities to optimize of our IR to the high level semantics in ATen. The responsibility for that naturally belongs to the lowerings in kernel.cpp and that included handling error conditions.

Reply:
I guess the only "dangerous" case that we can't check is when we have a load, in other cases we can probably assume that the IR was constructed by us and thus the divisor should not be zero.

Reply:
@ZolotukhinM I don't agree that's dangerous, that'd be a bug in the lowering. If we got in the position to divide by a load, we should have verified by that point that the loaded value wasn't zero. I don't believe we should make any effort to compensate for lowering bugs. Bugs need to be fixed where they are created.

Either way, I don't think we should hold this PR, let's remove the rule for now.

Reply:
It doesn't have to be a bug, check out the example I wrote in the comment above. This expression could come from the outside, we don't really have control over it. Note that it's relevant not only for RHS being 0, but also to LHS or RHS being NaN.

Agree on moving forward with this PR regardless.

Reply:
@ZolotukhinM I don't understand why we have no control over it. The right lowering of an `aten::div` would compare the divisor to 0, every single element of it, before generating the tensorexpr div. Maybe it sounds expensive, but this is the semantics.

Reply:
> In your example, we would first compare y against 0 and raise an error if it's zero. Then generate the tensorexpr lowering, which would subsequently be simplified by the rule added by this PR, which is correct when the non-zero pre-condition is met.

TBH, I don't think that would be feasible in general case. What if I have
```
def foo(x, y):
  return (x-y)/(x-y)
```
(and obviously I can construct even more complicated example)

Do you plan to first generate the expression for LHS and RHS (LHS to check for NaNs, RHS for 0s and NaNs), then generate compare vs 0, then generate the "usual" expression? I don't think it's better than just not simplifying in a presence of a load.

Reply:
Yes, that's how I'd lower it and this is how is done in all systems I'm aware of.

Reply:
> NNC can subsequently fuse those loops which do the checks and the division and still achieve the same results as if we baked the checks in the div itself.

NNC won't be able across potential exception throw (also, we don't have a way to represent a throw in TE at the moment).

I see where you're getting at, I just think it's too much work for what it achieves. What you suggest would be stricter in terms of semantics, but what we have now would also behave correctly unless we optimize this particular load case, and I think it might be just more practical to address it by not optimizing it.

Reply:
That's fine!

Reply:
That code enforces X / 0 = 0 in all paths though, it's DivNoNan, not Div.

I could see adding a DivNoNan operator though.

Reply:
@nickgg Fair enough, there is no exact equivalent between PyTorch and TF. PyTorch returns `inf` on division by 0 whereas TF returns `nan` iirc.

Not simplifying for now is fair enough, if we let it be we'll return `inf`, just as eager does.

Reply:
@asuhan does Pytorch return `inf` on integer div by 0? That's interesting. Without simplification I'd expect us to crash.

Reply:
@nickgg Yes, because division promotes to float. Integer (floor) division throws a `ZeroDivisionError`.